src/app/product.service.ts

```diff
@@ -0,0 +1,52 @@
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { map, publishReplay, refCount } from 'rxjs/operators';
+import { HttpClient } from '@angular/common/http';
+import { environment } from '../environments/environment';
+
+import { Product, SoldProducts } from './product';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductService {
+  private productsURL = `${environment.baseURL}/soldProducts`; 
+
+  productsData$?: Observable<SoldProducts>;
+
+  constructor(private http: HttpClient) { }
+
+  /** GET soldProducts from the server */
+  getSoldProducts(): Observable<SoldProducts> {
+    // Cache it once if productsData$ value is false
+    if (!this.productsData$) {
+      this.productsData$ = this.http.get<SoldProducts>(this.productsURL).pipe(
+        publishReplay(1),
+        refCount()
+      );
+    }
+    return this.productsData$;
+  }
+
+  getProgressValue(): Observable<number> {
+    return this.getSoldProducts().pipe(
+      map((productsData: SoldProducts) => {
+        const total: number = productsData.data.reduce((total:number, product:Product) => {
```

Comment:
Move this to calculate progress method as this have same responsibility (SOLID)

Reply:
Updated